File: 40-combination-sum-ii/combination-sum-ii.cpp
class Solution {
public:
    void solve(int idx, int n, vector<int>& candidates, int target,
               int sum, vector<int>& q, vector<vector<int>>& ans) {

        // base case
        if (sum == target) {
            ans.push_back(q);
            return;
        }
        if (idx >= n || sum > target) return;  // stop exploring

        // ✅ CHOOSE current element
        q.push_back(candidates[idx]);
        solve(idx + 1, n, candidates, target, sum + candidates[idx], q, ans);
        q.pop_back();

        // ✅ NOT CHOOSE current element
        // skip all duplicates of the current number before next recursion
        int nextIdx = idx + 1;
        while (nextIdx < n && candidates[nextIdx] == candidates[idx]) {
            nextIdx++;
        }
        solve(nextIdx, n, candidates, target, sum, q, ans);
    }

    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> ans;
        vector<int> q;
        solve(0, candidates.size(), candidates, target, 0, q, ans);
        return ans;
    }
};

File: 40-combination-sum-ii/combination-sum-ii.py
class Solution:
    def combinationSum2(self, nums: List[int], target: int) -> List[List[int]]:
        ansSet = set()
        nums.sort()
        def dfs(idx, curr, total):
            if total == target: 
                ansSet.add(tuple(curr.copy()))
                return 
            if total>target or idx>=len(nums):
                return 
            curr.append(nums[idx])
            dfs(idx+1, curr, total+nums[idx])
            curr.pop()
            nextIdx = idx+1
            while nextIdx<len(nums) and nums[idx]==nums[nextIdx]: nextIdx+=1
            dfs(nextIdx, curr, total)
        dfs(0,[],0)
        return list(ansSet)